From b7ddbd9c298e486aa168ef45485580fa6281d8cb Mon Sep 17 00:00:00 2001
From: Stephen Hemminger <sthemmin@microsoft.com>
Date: Tue, 14 Nov 2017 14:32:24 -0800
Subject: [PATCH 63/71] hv_netvsc: simplify function args in receive status
 path

The caller (netvsc_receive) already has the net device pointer,
and should just pass that to functions rather than the hyperv device.
This eliminates several impossible error paths in the process.

Signed-off-by: Stephen Hemminger <sthemmin@microsoft.com>
---
 drivers/net/hyperv/hyperv_net.h   |    3 +--
 drivers/net/hyperv/netvsc.c       |    2 +-
 drivers/net/hyperv/netvsc_drv.c   |   12 ++----------
 drivers/net/hyperv/rndis_filter.c |    9 +++------
 4 files changed, 7 insertions(+), 19 deletions(-)

--- a/drivers/net/hyperv/hyperv_net.h	2017-11-17 10:41:16.050637559 -0800
+++ b/drivers/net/hyperv/hyperv_net.h	2017-11-17 10:41:16.046637609 -0800
@@ -197,7 +197,7 @@ int netvsc_send(struct net_device *ndev,
 		struct rndis_message *rndis_msg,
 		struct hv_page_buffer *page_buffer,
 		struct sk_buff *skb);
-void netvsc_linkstatus_callback(struct hv_device *device_obj,
+void netvsc_linkstatus_callback(struct net_device *net,
 				struct rndis_message *resp);
 int netvsc_recv_callback(struct net_device *net,
 			 struct vmbus_channel *channel,
@@ -220,7 +220,6 @@ int rndis_filter_set_rss_param(struct rn
 			       const u8 *key);
 int rndis_filter_receive(struct net_device *ndev,
 			 struct netvsc_device *net_dev,
-			 struct hv_device *dev,
 			 struct vmbus_channel *channel,
 			 void *data, u32 buflen);
 
--- a/drivers/net/hyperv/netvsc.c	2017-11-17 10:41:16.050637559 -0800
+++ b/drivers/net/hyperv/netvsc.c	2017-11-17 10:41:16.046637609 -0800
@@ -1083,7 +1083,7 @@ static int netvsc_receive(struct net_dev
 		u32 buflen = vmxferpage_packet->ranges[i].byte_count;
 
 		/* Pass it to the upper layer */
-		status = rndis_filter_receive(ndev, net_device, device,
+		status = rndis_filter_receive(ndev, net_device,
 					      channel, data, buflen);
 
 		trace_netvsc_receive(ndev, q_idx, buflen);
--- a/drivers/net/hyperv/netvsc_drv.c	2017-11-17 10:41:16.050637559 -0800
+++ b/drivers/net/hyperv/netvsc_drv.c	2017-11-17 10:41:16.046637609 -0800
@@ -653,22 +653,14 @@ no_memory:
 /*
  * netvsc_linkstatus_callback - Link up/down notification
  */
-void netvsc_linkstatus_callback(struct hv_device *device_obj,
+void netvsc_linkstatus_callback(struct net_device *net,
 				struct rndis_message *resp)
 {
 	struct rndis_indicate_status *indicate = &resp->msg.indicate_status;
-	struct net_device *net;
-	struct net_device_context *ndev_ctx;
+	struct net_device_context *ndev_ctx = netdev_priv(net);
 	struct netvsc_reconfig *event;
 	unsigned long flags;
 
-	net = hv_get_drvdata(device_obj);
-
-	if (!net)
-		return;
-
-	ndev_ctx = netdev_priv(net);
-
 	/* Update the physical link speed when changing to another vSwitch */
 	if (indicate->status == RNDIS_STATUS_LINK_SPEED_CHANGE) {
 		u32 speed;
--- a/drivers/net/hyperv/rndis_filter.c	2017-11-17 10:41:16.050637559 -0800
+++ b/drivers/net/hyperv/rndis_filter.c	2017-11-17 10:41:16.046637609 -0800
@@ -134,11 +134,9 @@ static void put_rndis_request(struct rnd
 	kfree(req);
 }
 
-static void dump_rndis_message(struct hv_device *hv_dev,
+static void dump_rndis_message(struct net_device *netdev,
 			       const struct rndis_message *rndis_msg)
 {
-	struct net_device *netdev = hv_get_drvdata(hv_dev);
-
 	switch (rndis_msg->ndis_msg_type) {
 	case RNDIS_MSG_PACKET:
 		netdev_dbg(netdev, "RNDIS_MSG_PACKET (len %u, "
@@ -396,7 +394,6 @@ static int rndis_filter_receive_data(str
 
 int rndis_filter_receive(struct net_device *ndev,
 			 struct netvsc_device *net_dev,
-			 struct hv_device *dev,
 			 struct vmbus_channel *channel,
 			 void *data, u32 buflen)
 {
@@ -418,7 +415,7 @@ int rndis_filter_receive(struct net_devi
 	}
 
 	if (netif_msg_rx_status(net_device_ctx))
-		dump_rndis_message(dev, rndis_msg);
+		dump_rndis_message(ndev, rndis_msg);
 
 	switch (rndis_msg->ndis_msg_type) {
 	case RNDIS_MSG_PACKET:
@@ -433,7 +430,7 @@ int rndis_filter_receive(struct net_devi
 
 	case RNDIS_MSG_INDICATE:
 		/* notification msgs */
-		netvsc_linkstatus_callback(dev, rndis_msg);
+		netvsc_linkstatus_callback(ndev, rndis_msg);
 		break;
 	default:
 		netdev_err(ndev,
